homeassistant/components/environment_canada/sensor.py

```diff
@@ -0,0 +1,172 @@
+    'dewpoint': {'name': 'Dew Point',
+                 'unit': TEMP_CELSIUS},
+    'wind_chill': {'name': 'Wind Chill',
+                   'unit': TEMP_CELSIUS},
+    'humidex': {'name': 'Humidex',
+                'unit': TEMP_CELSIUS},
+    'pressure': {'name': 'Pressure',
+                 'unit': 'kPa'},
+    'tendency': {'name': 'Tendency'},
+    'humidity': {'name': 'Humidity',
+                 'unit': '%'},
+    'visibility': {'name': 'Visibility',
+                   'unit': 'km'},
+    'condition': {'name': 'Condition'},
+    'wind_speed': {'name': 'Wind Speed',
+                   'unit': 'km/h'},
+    'wind_gust': {'name': 'Wind Gust',
+                  'unit': 'km/h'},
+    'wind_dir': {'name': 'Wind Direction'},
+    'high_temp': {'name': 'High Temperature',
+                  'unit': TEMP_CELSIUS},
+    'low_temp': {'name': 'Low Temperature',
+                 'unit': TEMP_CELSIUS},
+    'pop': {'name': 'Chance of Precip.',
+            'unit': '%'},
+    'warnings': {'name': 'Warnings'},
+    'watches': {'name': 'Watches'},
+    'advisories': {'name': 'Advisories'},
+    'statements': {'name': 'Statements'},
+    'endings': {'name': 'Ended'}
+}
+
+
+def validate_station(station):
+    """Check that the station ID is well-formed."""
+    if station is None:
+        return
+    if not re.fullmatch(r'[A-Z]{2}/s0000\d{3}', station):
+        raise vol.error.Invalid('Station ID must be of the form "XX/s0000###"')
+    return station
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_STATION): validate_station,
+    vol.Inclusive(CONF_LATITUDE, 'latlon'): cv.latitude,
+    vol.Inclusive(CONF_LONGITUDE, 'latlon'): cv.longitude,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Environment Canada sensor."""
+    from env_canada import ECData
+
+    if config.get(CONF_STATION):
+        ec_data = ECData(station_id=config[CONF_STATION])
+    elif config.get(CONF_LATITUDE) and config.get(CONF_LONGITUDE):
+        ec_data = ECData(coordinates=(config[CONF_LATITUDE],
+                                      config[CONF_LONGITUDE]))
+    else:
+        ec_data = ECData(coordinates=(hass.config.latitude,
+                                      hass.config.longitude))
+
+    add_devices([ECSensor(sensor_type, ec_data, config.get(CONF_NAME))
+                 for sensor_type in config[CONF_MONITORED_CONDITIONS]])
+
+
+class ECSensor(Entity):
+    """Implementation of an Environment Canada sensor."""
+
+    def __init__(self, sensor_type, ec_data, platform_name):
+        """Initialize the sensor."""
+        self.sensor_type = sensor_type
+        self.ec_data = ec_data
+        self.platform_name = platform_name
+        self._state = None
+        self._attr = None
+        self.update()
```

Comment:
We shouldn't be updating the entities in their `__init__`; if you want an immediate update, pass `True` as the second parameter to `add_devices`.